search/Iterative_deepening_search_8_tile_puzzle.cpp

```diff
@@ -0,0 +1,197 @@
+     */
+    bool dls(Node *currNode, int depthLimit) {
+        if (currNode->getDepth() > depthLimit) return false;
+        visitedList.insert(currNode->getState());
+
+        if (currNode->getState() == goalState) {
+            printSolution(currNode);
+            return true;
+        }
+
+        const vector<vector<int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
+        for (const auto &dir : dirs) {
+            int newI = currNode->getBlankIndex().first + dir[0];
+            int newJ = currNode->getBlankIndex().second + dir[1];
+
+            if (!boundsOK(newI, newJ)) continue;
+
+            vector<vector<int>> newState = currNode->getState();
+            swap(newState[currNode->getBlankIndex().first][currNode->getBlankIndex().second], newState[newI][newJ]);
+            Node *newNode = new Node(newState, currNode, currNode->getDepth() + 1);
+
+            if (visitedList.find(newNode->getState()) == visitedList.end()) {
+                if (dls(newNode, depthLimit)) return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @brief Initiates the iterative deepening search to solve the puzzle.
+     */
+    void iterativeDeepeningSearch() {
+        for (int depth = 0; depth < INT_MAX; depth++) {
+            visitedList.clear();
+            Node *root = new Node(initialState, nullptr, 0);
+            if (dls(root, depth)) return;
+        }
+    }
+
+    /**
+     * @brief Prints the solution path.
+     * @param node The node at which the solution is found.
+     */
+    void printSolution(Node *node) {
+        cout << "Solution found at depth " << node->getDepth() << endl;
+        // Trace back to print the solution path if necessary
+    }
+
+    /**
+     * @brief Checks if the puzzle is solvable based on inversions.
+     * @return True if solvable, false otherwise.
+     */
+    bool isSolvable() const {
+        int inversions = 0;
+        for (int i = 0; i < N * N; i++) {
+            for (int j = i + 1; j < N * N; j++) {
+                if (initialState[i / N][i % N] && initialState[j / N][j % N] &&
+                    initialState[i / N][i % N] > initialState[j / N][j % N]) {
+                    inversions++;
+                }
+            }
+        }
+        return (inversions % 2 == 0);
+    }
+};
+
+/**
+ * @brief Takes user input for the initial state of the puzzle.
+ * @param initialState Reference to the 2D vector where the input will be stored.
+ */
+void takeUserInput(vector<vector<int>> &initialState) {
+    cout << "Enter the initial state of the 8-tile puzzle (0 for blank tile):\n";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            cin >> initialState[i][j];
+        }
+    }
+}
+
+/**
+ * @brief Main function to run the 8-tile puzzle solver.
+ */
+int main() {
+    vector<vector<int>> initialState(N, vector<int>(N, 0));
+    vector<vector<int>> goalState = {{1, 2, 3}, {4, 5, 6}, {7, 8, 0}};
+
+    takeUserInput(initialState);
+
```

Comment:
Do not have anything else in the main function other than a call to a test function

Reply:
did changes in recent PR